### DISCUSSION TITLE
Overreding evcxr rust compiler

### DISCUSSION THREAD

**Discussion by iazel** [ipetkov/crane#415] [Q&A]
Hello, I'm struggling with overriding the rust implementation for evcxr package, the terminal REPL for rust.

As far as I can see in [evcxr](https://github.com/NixOS/nixpkgs/blob/4467663aaa4d20b329e53ebe631bd3092497c527/pkgs/development/interpreters/evcxr/default.nix) nix file, it expects a `rustPlatform`, which we could make using [pkgs.makeRustPlatform](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/compilers/rust/make-rust-platform.nix), however it seems like the `rustc` this one expects isn't compatbile with our `rust-bin.stable.latest.rustc` nor `rust-bin.stable.latest.default`, so I don't understand how I could make it work.

Any suggestion would be highly appreciated! :pray: 

This is my current attempt:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs";
    flake-utils.url = "github:numtide/flake-utils";
    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
      inputs.flake-utils.follows = "flake-utils";
    };
  };

  outputs = { self, nixpkgs, flake-utils, rust-overlay }:

    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs { 
          inherit system;
          overlays = [ (import rust-overlay) ];
        };
        rust = pkgs.rust-bin.stable.latest.default.override {
          extensions = [
            "rust-src"
            "rust-analysis"
            ];
        };
        evcxr = pkgs.evcxr.override { 
            rustPlatform = pkgs.makeRustPlatform { 
                rustc = rust;
                cargo = rust.cargo;
            };
        };
      in with pkgs;
      {
        devShells.default = mkShell {
          buildInputs = [ pkg-config rust rust-analyzer cargo-watch lldb evcxr ];
        };
      }
    );
}
```

Which produces this error:
```
error:
       … while calling the 'derivationStrict' builtin

         at //builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'nix-shell'
         whose name attribute is located at /nix/store/8167w2s280vs6wkcnmjb2zl8082jlzsp-source/pkgs/stdenv/generic
/make-derivation.nix:300:7

       … while evaluating attribute 'buildInputs' of derivation 'nix-shell'

         at /nix/store/8167w2s280vs6wkcnmjb2zl8082jlzsp-source/pkgs/stdenv/generic/make-derivation.nix:347:7:

          346|       depsHostHost                = lib.elemAt (lib.elemAt dependencies 1) 0;
          347|       buildInputs                 = lib.elemAt (lib.elemAt dependencies 1) 1;
             |       ^
          348|       depsTargetTarget            = lib.elemAt (lib.elemAt dependencies 2) 0;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'src' missing

       at /nix/store/8167w2s280vs6wkcnmjb2zl8082jlzsp-source/pkgs/development/compilers/rust/rust-lib-src.nix:4:35
:

            3| runCommand "rust-lib-src" { } ''
            4|   tar --strip-components=1 -xzf ${rustc.src}
             |                                   ^
            5|   mv library $out
```

**Comment by ipetkov** [comment]
Hi @iazel !

The way `rustc` is packaged into nixpkgs differs somewhat from what `oxalica/rust-overlay` is able to provide. The error you are seeing [comes from here](https://github.com/NixOS/nixpkgs/blob/2d386760f73f26892c0dd9cacaeef0f11d99246b/pkgs/development/compilers/rust/rust-lib-src.nix#L4), specifically it assumes that the `rustc` object has an attribute called `src` (which `oxalica/rust-overlay`'s version does not).

Your current example is trying to replace the rust toolchain of a derivation built with `rustPlatform` from nixpkgs, so you may need to patch how nixpkgs's version of `rustc` is compiled. Alternatively you could write a new definition for this package (using `crane` or whatever else you wish) and side step that altogether (we've got a number of examples if you'd like a starting point!)

Otherwise I'm afraid I don't know enough about the nixpkgs internals on how to suggest fixing the issue, but you may also want to try stopping by the NixOS community matrix channel and ask there.

Good luck!

**Reply by iazel** [reply]
Hi Ivan, thank you for all the information and suggestion :+1:
